crates/gen-host-js/src/lib.rs

```diff
@@ -418,69 +423,89 @@ impl Js {
 
             Intrinsic::I32ToF32 => self.src.js("
                 function i32ToF32(i) {
-                    I32_TO_F32_I[0] = i;
-                    return I32_TO_F32_F[0];
+                    i32ToF32I[0] = i;
+                    return i32ToF32F[0];
                 }
             "),
             Intrinsic::F32ToI32 => self.src.js("
                 function f32ToI32(f) {
-                    I32_TO_F32_F[0] = f;
-                    return I32_TO_F32_I[0];
+                    i32ToF32F[0] = f;
+                    return i32ToF32I[0];
                 }
             "),
             Intrinsic::I64ToF64 => self.src.js("
                 function i64ToF64(i) {
-                    I64_TO_F64_I[0] = i;
-                    return I64_TO_F64_F[0];
+                    i64ToF64I[0] = i;
+                    return i64ToF64F[0];
                 }
             "),
             Intrinsic::F64ToI64 => self.src.js("
                 function f64ToI64(f) {
-                    I64_TO_F64_F[0] = f;
-                    return I64_TO_F64_I[0];
+                    i64ToF64F[0] = f;
+                    return i64ToF64I[0];
                 }
             "),
 
             Intrinsic::Utf8Decoder => self
                 .src
-                .js("const UTF8_DECODER = new TextDecoder('utf-8');\n"),
+                .js("const utf8Decoder = new TextDecoder('utf-8');\n"),
 
-            Intrinsic::Utf8EncodedLen => self.src.js("let UTF8_ENCODED_LEN = 0;\n"),
+            Intrinsic::Utf16Decoder => self
+                .src
+                .js("const utf16Decoder = new TextDecoder(`utf-16${isLE ? '' : 'be'}`);\n"),
```

Comment:
This was somewhat surprising to me in that I would expect this to always use `utf-16le` as the encoding since wasm is always little-endian. I get mixed up with endianness though so I'm not sure if the `be` is to swap things back to normal or not.

Reply:
`utf-16` means `utf-16le` per the spec, so that's why I left out the explicit `le`. I think `new TextDecoder()` without any arg is supposed to work with a Uint16Array as UTF16 and it seems to, I was just unsure if that mode would try to revert to UTF8 somehow so was being explicit. Happy to adjust further.

Reply:
> I think new TextDecoder() without any arg is supposed to work with a Uint16Array as UTF16 and it seems to,

Locally on Firefox I get:

```
(new TextDecoder()).decode(new Uint16Array([0, 0])) == "\u0000\u0000\u0000\u0000" 
(new TextDecoder('utf-16')).decode(new Uint16Array([0, 0])) == "\u0000\u0000" 
```

so I think that an argument is required here?

> utf-16 means utf-16le per the spec

Ah I was not aware of that, I mistakenly presumed `utf-16` was utf-16 with native-endianness. In that case `utf-16` should be sufficient to always pass here.

I don't believe it's correct to use `utf-16be` here. When I run the tests in `qemu-s390x` the tests fail, and when I remove the `be` here the tests pass.

Reply:
> I don't believe it's correct to use utf-16be here. When I run the tests in qemu-s390x the tests fail, and when I remove the be here the tests pass.

I'm pretty sure according to the specs then that what we have currently for big endian handling is correct but that there may be another big endian related failure happening in the code perhaps due to the other intrinsics?

Reply:
Perhaps yeah? We don't test big-endian in CI but if you're looking to handle endianness here I think it would be good to test at least locally that the endianness is right (for me at least this PR as-is does not pass tests)

Reply:
I really think it's very important we sort out endianness here, so removing logic that works towards that goal seems counter productive to me.

Reply:
The techniques I'm using have been tested before on big endian hardware, even if unrelated bindgen support is likely causing further failures.

Reply:
Per further discussion, the decoder is now using UTF16 LE due to the underlying buffer being a Wasm little endian buffer.